src/Workspaces/Core/Portable/Execution/AbstractReferenceSerializationService.cs

```diff
@@ -518,10 +518,13 @@ private void WriteTo(AnalyzerReference reference, ObjectWriter writer, bool chec
                 writer.WriteString(nameof(AnalyzerFileReference));
                 writer.WriteInt32((int)SerializationKinds.FilePath);
 
-                if (!checksum)
+                if (checksum)
                 {
                     // we don't write full path when creating checksum
-                    writer.WriteString(file.FullPath);
+                    // make sure we always normalize assemblyPath to lower case since it comes from Assembly.Location
+                    // unlike FullPath which comes from string
+                    writer.WriteString(assemblyPath.ToLowerInvariant());
```

Comment:
This is not compatible with FileSystems (including NTFS on Windows) that allow files to differ by case.